### PR TITLE
[CHK-839] alternative business logic for payment gateway invocation

### PR DESCRIPTION
**Pull request by ciuffagianluca**
Refinement for payment gateway invocation

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java [hunk 1]
```diff
@@ -47,47 +49,41 @@ public Mono<RequestAuthorizationResponseDto> handle(TransactionRequestAuthorizat
             return Mono.error(new AlreadyProcessedException(transaction.getRptId()));
         }
 
-        return paymentGatewayClient.requestGeneralAuthorization(command.getData())
-                .flatMap(authResponse -> {
-                    log.info("Logging authorization event for rpt id {}", transaction.getRptId().value());
+        var monoPostePay = Mono.just(command.getData())
+                .flatMap(d -> paymentGatewayClient.requestPostepayAuthorization(d))
+                .switchIfEmpty(Mono.empty())
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java:54]: Redundant?

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java [hunk 2]
```diff
@@ -47,47 +49,41 @@ public Mono<RequestAuthorizationResponseDto> handle(TransactionRequestAuthorizat
...
+                .map(p -> Tuples.of(p.getRequestId(),p.getUrlRedirect()));
 
-                    if(!authResponse.getT1().isPresent() && !authResponse.getT2().isPresent()) {
-                        return Mono.error(new BadRequestException("No gateway matched"));
-                    }
+        var monoXPay = Mono.just(command.getData())
+                .flatMap(d -> paymentGatewayClient.requestXPayAuthorization(d))
+                .switchIfEmpty(Mono.empty())
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java:59]: Also here :)

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java [hunk 3]
```diff
@@ -47,47 +49,41 @@ public Mono<RequestAuthorizationResponseDto> handle(TransactionRequestAuthorizat
...
+                .map(p -> Tuples.of(p.getRequestId(),p.getUrlRedirect()));
 
-                    Mono<Tuple2<String,String>> monoPostePay = Mono.just(authResponse.getT1()).filter(pPayAuth -> pPayAuth.isPresent())
-                            .switchIfEmpty(Mono.empty())
-                            .flatMap(pPay -> Mono.zip(Mono.just(pPay.get().getRequestId()),Mono.just(pPay.get().getUrlRedirect())));
+        return monoPostePay.switchIfEmpty(monoXPay).switchIfEmpty(Mono.error(new BadRequestException("No gateway matched")))
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java:62]: What about creating a list of `Mono`s and reducing over that?
Something like:

```java
List<Mono<Tuple<String, String>>> gatewayRequests = List.of(monoPostePay, monoXPay);

Mono<Tuple2<String, String>> gatewayAttempts = gatewayRequests.stream().reduce((pipeline, candidateStep) -> {
    return pipeline.switchIfEmpty(candidateStep);
});

gatewayAttempts.switchIfEmpty(Mono.error(new BadRequestException("No gateway matched")));
```

File: src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java [hunk 4]
```diff
@@ -132,8 +136,12 @@ void shouldReturnAuthorizationResponseCP_XPAY() throws JsonProcessingException {
                 .thenReturn(Mono.just(xPayResponse));
 
         /* test */
-        StepVerifier.create(client.requestGeneralAuthorization(authorizationData))
-                .assertNext(response -> response.getT2().get().equals(xPayResponse))
+        StepVerifier.create(client.requestXPayAuthorization(authorizationData))
+                .assertNext(response -> response.equals(xPayResponse))
```
Comment (giovanniberti) [inline src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java:140]: If you use `expectNext` you get nicer console output (also watch out, `assertNext` expects you to throw assertion errors inside the lambda body, `expectNext` does that for you)

```suggestion
                .expectNext(xPayResponse)
```

File: src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java [hunk 5]
```diff
@@ -181,8 +189,12 @@ void shouldReturnAuthorizationResponsePPAY() throws JsonProcessingException {
...
 
         /* test */
-        StepVerifier.create(client.requestGeneralAuthorization(authorizationData))
-                .assertNext(response -> response.getT1().get().equals(Mono.just(postePayResponse)))
+        StepVerifier.create(client.requestXPayAuthorization(authorizationData))
+                .expectNextCount(0)
+                .verifyComplete();
+
+        StepVerifier.create(client.requestPostepayAuthorization(authorizationData))
+                .expectNextMatches(t -> t.equals(postePayResponse))
```
Comment (giovanniberti) [inline src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java:197]:
```suggestion
                .expectNext(postePayResponse)
```

File: src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java [hunk 6]
```diff
@@ -466,8 +499,12 @@ void fallbackOnEmptyMdcInfoOnMapperErrorPPAY() throws JsonProcessingException {
                 .thenReturn(Mono.just(postePayResponse));
 
         /* test */
-        StepVerifier.create(client.requestGeneralAuthorization(authorizationData))
-                .assertNext(response -> response.getT1().get().equals(postePayResponse))
+        StepVerifier.create(client.requestPostepayAuthorization(authorizationData))
+                .expectNextMatches(response -> response.equals(postePayResponse))
```
Comment (giovanniberti) [inline src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java:503]:
```suggestion
                .expectNext(postePayResponse)
```

File: src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java [hunk 7]
```diff
@@ -518,8 +555,12 @@ void fallbackOnEmptyMdcInfoOnMapperErrorCP_XPAY() throws JsonProcessingException
                 .thenReturn(Mono.just(xPayResponse));
 
         /* test */
-        StepVerifier.create(client.requestGeneralAuthorization(authorizationData))
-                .assertNext(response -> response.getT2().get().equals(xPayResponse))
+        StepVerifier.create(client.requestXPayAuthorization(authorizationData))
+                .expectNextMatches(response -> response.equals(xPayResponse))
```
Comment (giovanniberti) [inline src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java:559]:
```suggestion
                .expectNext(xPayResponse)
```

**Review by giovanniberti — COMMENTED** [review]
Good job! Left a few minor comments 🚀

File: src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java (hunk 3 above)
Comment (ciuffagianluca) [reply src/main/java/it/pagopa/transactions/commands/handlers/TransactionRequestAuthorizationHandler.java:62]: That's a code emotion 💘

File: src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java [hunk 8]
```diff
@@ -227,7 +239,12 @@ void shouldThrowAlreadyProcessedOn401CP_XPAY() throws JsonProcessingException {
                 .thenReturn(Mono.error(new WebClientResponseException("api error", HttpStatus.UNAUTHORIZED.value(), "Unauthorized", null, null, null)));
 
         /* test */
-        StepVerifier.create(client.requestGeneralAuthorization(authorizationData))
+
```
Comment (giovanniberti) [inline src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java:242]: You could make these checks even stronger by adding a `Mockito.verify(client, Mockito.times(0)).requestPostepayAuthorization(authorizationData)`: the assertion you put checks that indeed `PaymentGatewayClient` outputs an empty `Mono`, with the other one you also check that the postepay authorization method is never called 🙂 (see also other instances of this)

Comment (giovanniberti) [reply src/test/java/it/pagopa/transactions/client/PaymentGatewayClientTest.java:315]: Solved in [5c3afec](https://github.com/pagopa/pagopa-ecommerce-transactions-service/pull/92/commits/5c3afecb6aff7bbdefdbe1defb58fbc5dd79b455)

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**